### ISSUE TITLE
Linter not working on Windows

### ISSUE THREAD

**Issue by AresEkb** [isabelle-prover/isabelle-linter#18]
Should I copy `jedit_linter` folder to `$ISABELLE_HOME/contrib` and run the following command (on Windows)?

```
Isabelle2021-1.exe components -u jedit_linter
```

How to check that Linter is installed? It seems that nothing is changed.

**Comment by TheMC47** [comment]
I think the `Isabelle2021-1.exe` executable is different from the `/bin/isabelle` executable, as far as I understand. You need to run the `isabelle` binary from the `bin` directory (e.g. `PATH_TO_ISABELLE_CLONE/bin/isabelle components -u LINTER_DIR`). Oh, and I guess you need to launch jedit with `PATH_TO_ISABELLE_CLONE/bin/isabelle jedit`

I hope that works for now, not sure if there's a better way

**Comment by AresEkb** [comment]
I've got it. I did the following steps:

1. Copy `jedit_linter` folder to `contrib` folder
2. Run `Cygwin-Terminal.bat`
3. Run `isabelle components -u contrib/jedit_linter`
4. Run `isabelle jedit`

I get the following errors:
```
### Building Isabelle/Linter (C:\Tools\Isabelle2021-1\lib\classes\isabelle_linter_plugin.jar) ...
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\jedit_extension.scala:8: error: object linter is not a member of package isabelle
import isabelle.linter._
                ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\jedit_extension.scala:33: error: not found: value Linter_Markup
    val ext_elements = Markup.Elements(Linter_Markup.LINTER_SENDBACK, Linter_Markup.GOTO_POSITION)
                                       ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\jedit_extension.scala:33: error: not found: value Linter_Markup
    val ext_elements = Markup.Elements(Linter_Markup.LINTER_SENDBACK, Linter_Markup.GOTO_POSITION)
                                                                      ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\jedit_extension.scala:73: error: not found: value Linter_Markup
        case XML.Elem(Markup(Linter_Markup.LINTER_SENDBACK, props), _) =>
                             ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\jedit_extension.scala:79: error: not found: value Linter_Markup
        case XML.Elem(Markup(Linter_Markup.GOTO_POSITION, pos), _) =>
                             ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:6: error: object linter is not a member of package isabelle
import isabelle.linter._
                ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:16: error: not found: type Linter_Variable
class PIDE_Linter_Variable extends Linter_Variable(true)
                                   ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:16: error: no arguments allowed for nullary constructor Object: (): Object
class PIDE_Linter_Variable extends Linter_Variable(true)
                                                   ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:21: error: not found: value get
      linter <- get
                ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\overlay_reporter.scala:5: error: object linter is not a member of package isabelle
import isabelle.linter._
                ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\overlay_reporter.scala:8: error: not found: type Reporter
object Overlay_Reporter extends Reporter[Linter_Overlay.State]
                                ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\overlay_reporter.scala:16: error: not found: value Linter
  override def report_for_snapshot(lint_report: Linter.Lint_Report, show_desriptions: Boolean = false): Linter_Overlay.State =
                                                ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:77: error: not found: value XML_Reporter
  val separator = XML_Reporter.text("----------------")
                  ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:78: error: not found: value XML_Reporter
  val disabled = XML_Reporter.text("The linter plugin is disabled.")
                 ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:88: error: value get is not a member of isabelle.jedit_linter.PIDE_Linter_Variable
      val new_output = Linter_Plugin.instance.flatMap(_.linter.get) match {
                                                               ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:94: error: value lint_report is not a member of Nothing
          val report = linter.lint_report(snapshot)
                              ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:95: error: not found: value XML_Reporter
          val snapshot_lints = XML_Reporter.report_for_snapshot(report, show_descriptions)
                               ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:96: error: not found: value XML_Reporter
          val command_lints = XML_Reporter.report_for_command(report, current_command.id)
                              ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_dockable.scala:179: error: value update is not a member of isabelle.jedit_linter.PIDE_Linter_Variable
        plugin.linter.update(PIDE.options.value)
                      ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_plugin.scala:37: error: value update is not a member of isabelle.jedit_linter.PIDE_Linter_Variable
          linter.update(PIDE.options.value)
                 ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\linter_plugin.scala:46: error: value update is not a member of isabelle.jedit_linter.PIDE_Linter_Variable
          linter.update(PIDE.options.value)
                 ^
C:\Tools\Isabelle2021-1\contrib\jedit_linter\src\overlay_reporter.scala:10: error: not found: value Linter
  override def report_for_command(lint_report: Linter.Lint_Report,
                                               ^
22 errors
*** Failed to compile Scala sources
```

Should it work on Windows?

**Comment by TheMC47** [comment]
`jedit_linter` is not enough, you need to add the whole directory (e.g. `isabelle components -u DIR`, where  `DIR` is the cloned linter directory

**Comment by AresEkb** [comment]
Thanks! I've build it, but now I get the following error:

```
$ isabelle jedit
### Building Isabelle Linter (C:\Tools\Isabelle2021-1\lib\classes\isabelle_linter.jar) ...
### Building Isabelle/Linter (C:\Tools\Isabelle2021-1\lib\classes\isabelle_linter_plugin.jar) ...
5:00:18 PM [main] [error] PluginJAR: Error while starting plugin isabelle.jedit_linter_plugin.Plugin
5:00:19 PM [main] [error] PluginJAR: java.lang.reflect.InvocationTargetException
5:00:19 PM [main] [error] PluginJAR:  at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
5:00:19 PM [main] [error] PluginJAR:  at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
5:00:19 PM [main] [error] PluginJAR:  at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
5:00:19 PM [main] [error] PluginJAR:  at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
5:00:19 PM [main] [error] PluginJAR:  at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
5:00:19 PM [main] [error] PluginJAR:  at org.gjt.sp.jedit.PluginJAR.activatePlugin(PluginJAR.java:938)
5:00:19 PM [main] [error] PluginJAR:  at org.gjt.sp.jedit.PluginJAR.activatePluginIfNecessary(PluginJAR.java:1026)
5:00:19 PM [main] [error] PluginJAR:  at org.gjt.sp.jedit.jEdit.main(jEdit.java:530)
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit.Main$.$anonfun$main$5(main.scala:130)
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit.Main$.main(main.scala:139)
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit.Main.main(main.scala)
5:00:19 PM [main] [error] PluginJAR: Caused by: java.lang.NoClassDefFoundError: isabelle/jedit_linter/jedit_extension
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit_linter.JEdit_Extension$.remove_final(jedit_extension.scala:27)
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit_linter.JEdit_Extension$.init(jedit_extension.scala:36)
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit_linter.Linter_Plugin.<init>(linter_plugin.scala:12)
5:00:19 PM [main] [error] PluginJAR:  at isabelle.jedit_linter_plugin.Plugin.<init>(plugin.scala:6)
5:00:19 PM [main] [error] PluginJAR:  ... 11 more
5:00:19 PM [main] [error] PluginJAR: Caused by: java.lang.ClassNotFoundException: isabelle.jedit_linter.jedit_extension
5:00:19 PM [main] [error] PluginJAR:  at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
5:00:19 PM [main] [error] PluginJAR:  at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
5:00:19 PM [main] [error] PluginJAR:  at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
5:00:19 PM [main] [error] PluginJAR:  ... 15 more
5:00:19 PM [main] [error] ErrorListDialog$ErrorEntry: C:\Users\Denis\.isabelle\Isabelle2021-1\jedit\jars\isabelle_jedit_linter.jar:
5:00:19 PM [main] [error] ErrorListDialog$ErrorEntry: Cannot start: java.lang.reflect.InvocationTargetException
```

**Comment by TheMC47** [comment]
Can you undo "Copy jedit_linter folder to contrib folder" and try again? Don't copy anything manually to the folders

**Comment by AresEkb** [comment]
Yes, I re-installed Isabelle before previous attempt. I've got that Linter is installed into user home folder and should not be copied manually.

**Comment by TheMC47** [comment]
I'll try to reproduce it on windows today

**Comment by TheMC47** [comment]
I can reproduce the error, these are the steps:

1. Download the Windows executable from https://isabelle.in.tum.de/index.html and run it
2. Download the linter source (in the directory LINTER_DIR)
3. Open `Cygwin-Terminal`
4. Type `isabelle components -u LINTER_DIR` 
5. Run `isabelle -?` to verify that the linter is installed. The tool `lint` should be displayed
6. Run `isabelle jedit`, the error described above show sup

One thing to note is that when running the executable created during installation, jedit opens and the linter plugin options are visible. The linter *doesn't* work there though.

I'll need to look further into that. @Dacit  do you have an idea why this might happen?

**Comment by Dacit** [comment]
Fixed by da990f769adbf6a41af465acfbf55829ecf6d4cd
